### ISSUE TITLE
The in-memory mode drive="core"  ,  is it not working?

### ISSUE THREAD

**Issue by jishin86** [h5py/h5py#2613]
Hello everyone!

I’m working with an NFS-mounted directory containing tens of thousands of JPG patch images for training an AI model, but file access is too slow (spend 8 hours to read whole files). To address this, I attempted to load the entire dataset into memory using the driver="core" option in h5py.

I created an HDF5 file where each group replicates the original JPG file path structure, and each key corresponds to a file path, storing the binary content of each image as a NumPy array. Saving and reading this HDF5 file worked without any issues in my initial tests.

However, when reading the HDF5 file using driver="core" and converting the binary arrays back into images, I noticed that the access speed was actually slower than reading the JPG files directly from HDD. (spend 12 hours to read whole file from hdf5 in memory)

This suggests that the in-memory functionality may not be working as expected. Am I missing something?

in my code, I read hdf5 file by
    f=h5py.File(h5_path, 'r', driver=driver)
And I saw the memory usage increment as much as HDF5 file size (30GB)

and read image  in memory
    img = f[img_path]
    bin_img = np.array(img)
    img_res = Image.open(io.BytesIO(bin_img))

I saw the offical example at
https://docs.h5py.org/en/stable/high/file.html#in-memory-files
And I tested it in same way 

    f=h5py.File(h5_path, 'r')
    hdf_data = f.id.get_file_image()
    f2 = h5py.File.in_memory(hdf_data)

but it was same result (time consuming)

All case was test in single thread (as pytorch dataloader single worker)

**Comment by takluyver** [comment]
I believe the "core" driver works as intended, though that's really up to HDF5, h5py just passes through the setting to use that driver.

Performance is always hard to guess about, though. From your description, it sounds like you're maybe storing the images in the HDF5 file uncompressed, which probably means a lot more data to read than the JPEG files.

12 hours for 30 GB is definitely slow, however you look at it. I'd suggest you start by working out how fast you can read data over the network with no parsing of any kind, just seeing how long it takes to read a GB. This might be a bit tricky if the filesystem does some caching, but there are ways round that. If you're accessing the data over a slow network link, fiddling with the format and the code isn't going to help (although using a few workers might make it a bit quicker 🤷 ).

**Comment by jishin86** [comment]
@takluyver 

Thank you for your comments.

The 30GB of data refers to a collection of compressed JPG image files. In the HDF5 file, I stored the binary contents of each JPG file directly, and at read time, I decode them into NumPy arrays using np.array(Image.open(...)). The goal was to reuse the original codebase, which reads individual JPG files from disk, without modification.
I followed the approach described here:

https://github.com/h5py/h5py/issues/745

My expectation was that, compared to reading many small JPG files from a network-mounted HDD and decoding them one by one, reading the same binary blobs from an in-memory HDF5 container and decoding them entirely in RAM would be significantly faster.

However, I found that even on my local PC with an SSD, whether I use the core driver or not, reading the entire 30GB HDF5 file still takes about 4 hours. This raises the question of whether the performance bottleneck lies in the code itself, or if the SSD is already fast enough that in-memory reading offers no real advantage.

More importantly, I observed that when running the same process on the server, the in-memory HDF5 approach using the core driver actually takes significantly longer than on my local machine, which is unexpected. I'm currently trying to understand whether the delay is due to server-side I/O performance, memory constraints, or something in the code logic itself.

**Comment by jhendersonHDF** [comment]
Hi @jishin86,

https://github.com/HDFGroup/hdf5/issues/3275 may be of interest, especially if you're using Windows. Though, even outside of windows there are some improvements we could probably make to the way that the core driver requests more memory as needed.

**Comment by Delengowski** [comment]
I'll say I work exclusively on a remote RHEL servers with NFS and when I was doing work to make read time improvements through different drivers and chunking strategies. I saw zero difference at all between the core and default driver - I found this odd. In retrospect I wish I did some memory profiling to observe what's happening. I believe I chalked it up to most read time occuring while decompressing.

**Comment by takluyver** [comment]
Closing this as I don't think there's any concrete action for h5py. As far as we know, the core driver does work. If you see poor performance, profile & experiment to understand what parts are slow. Feel free to open new issues if there's concrete reason to suspect that h5py is doing something wrong - or use [the HDF forum](https://forum.hdfgroup.org/) to discuss HDF5 performance more generally.
